[#141] 글쓰기: 사진 선택 화면 구현

jwonyLee opened the pull request:
## 개요

<img src="https://user-images.githubusercontent.com/15073405/139302270-77364e92-0488-416a-a18f-c7232e80c8cb.gif" width="50%" />

- 사진 선택 화면 구현
- 사용자 앨범 목록 화면 구현
- 선택한 이미지 썸네일 표시 구현

## 작업 사항

- 사용자 사진 권한 요청 기능 구현 (권한 거부 시, 설정으로 이동하는 Alert 표시)
- `PhotosViewController` (선택한 앨범 목록 화면, 기본값: 전체)
    - 네비게이션 - 확인 버튼은 선택 항목이 있으면 활성화, 없으면 비활성화
    - 네비게이션 - 확인 버튼을 누르면 글쓰기 화면으로 돌아가면서 선택한 이미지 데이터 전달 (NotificationCenter)
    - 사진 선택 기능 구현 (10개 제한, 초과시 Alert 표시)
- `AlbumsViewController` (사용자 앨범 목록 화면, 사진만 표시 (동영상 제외))
    - 선택한 앨범 전달, `PhotosViewController` 에서 해당 앨범만 표시 (이 때 기존에 선택했던 사진 목록은 초기화됨, 병연님이랑 협의된 부분) (NotificationCenter)
    - `ImageWithVerticalTwoLabelTableCell` 구현
        - `AlbumsViewController` 에서 앨범을 표시하기 위한 셀
- `WriteViewController`
    - `PhotosViewController`에서 전달받은 이미지들을 `ThumbnailCell`에 이미지 적용
    - `ThumbnailCell.closeButton` 선택하면 썸네일 이미지 제거 기능 구현
    - `ThumbnailImages` 개수에 따라 `CameraButtonCollectionCell` 업데이트

## 예외 사항

- 선택한 앨범 목록 화면에서 카메라 기능 미구현
    - 카메라 셀은 카메라 기능 추가하면서 바꿀 예정입니다.
- 선택한 앨범 목록 화면에서 단일 이미지 선택 기능 미구현

## 버그

- `AlbumsViewController`를 표시할 때 간혈적으로 튀는 현상

### Linked Issue

close #141

jwonyLee committed:
[#141] feat: CameraButtonCollectionCell 터치하면 PhotosViewController 표시

jwonyLee committed:
[#141] feat: 사진 권한 요청, 사진 선택 화면 구현
- ContentsCollectionViewCell 일부 수정
- 사용자에게 사진 접근 권한 요청 기능 구현
- 권한이 거부당하면 설정으로 이동하는 Alert 구현
- 권한이 허용되면 PhotosViewController로 이동, 사용자 앨범에 있는 모든 사진 표시

jwonyLee committed:
[#141] feat: 사진 선택 기능 구현
- 10개 이상 사진 선택 시 사용자에게 Alert을 보여준다.
- 중간에 선택한 항목이 변경되면 선택되어있던 셀들을 업데이트한다.
- 선택한 항목 개수만큼 SuccessButton 의 숫자를 변경한다.

jwonyLee committed:
[#141] feat: AlbumsViewController 구성
- ImageWithVerticalTwoLabelTableCell 구현
- 사용자 앨범 표시

jwonyLee committed:
[#141] feat: 앨범 선택 시 PhotosViewController 에서 선택한 앨범 표시

jwonyLee committed:
[#141] view: 선택한 사진이 없으면 확인 버튼 비활성화

jwonyLee committed:
[#141] ref: 코드 파일 분리

jwonyLee committed:
[#141] ref: 이미지 우측 상단 클릭 시만 다중 선택 가능하게 수정

jwonyLee committed:
[#141] feat: 선택한 이미지를 글쓰기 화면으로 전달

jwonyLee committed:
[#141] view: 선택한 사진 썸네일 사진 표시

gustn3965 reviewed (commented):
우선적으로 동작화면들을 보았습니다.  자세한 코드는 다시 확인할게요~ 

1.풀리퀘 올리기 전에 폴더내의 파일들 정렬해서 올려주세요 
    - ![image](https://user-images.githubusercontent.com/48749182/139387392-0afd6027-eba8-4ec3-a8b7-b99b16941e79.png)


2. `PhotosViewController`에서 사진 선택하다보면 탭바 우측 카운팅 되면서 레이블 크기가 계속 달라지는 것 같습니다
    -  320크기의 디바이스에서는 짤리는 경우도 있습니다. 
    - ![Pasted Graphic 4](https://user-images.githubusercontent.com/48749182/139386922-b37713e5-b22c-41f4-82db-cc1a7226d344.jpg)


3. `PhotosViewController`띄우고, 다시 사진첩선택에서 어떤 앨범 선택하고 다시 `PhotosViewController`로 온 경우, 첫번째 셀에 동그라미 체크 가 들어가는 위치를 탭하면 카운팅됩니다. 


5. iOS 14, 14.5 에서 PhotosViewController의 첫번째 셀 이미지가 나타나지 않습니다.
    - ![Pasted Graphic 5](https://user-images.githubusercontent.com/48749182/139386998-e54ca751-db29-41ac-ad29-90cde5c10a01.jpg)



6.이미지들이 셀안에서 셀크기에 맞게 늘어지기 때문에 다른 mode가 필요할 것 같습니다 ( `PhotosViewController`,  `WriteViewController` )



7. 잘 지키시는데, 최대한 메소드 위치들을 특정 기준에 맞게 나누어주셨으면 좋겠습니다. 
    - ![WriteViewController](https://user-images.githubusercontent.com/48749182/139387068-684acbb4-e7e0-420f-b65a-876cec5166e8.jpg)


8. SwiftLint의 워닝들은 최대한 지켜주시면 좋겠습니다 ( import 순서,  파라미터 라인 정렬 ) 
![image](https://user-images.githubusercontent.com/48749182/139387326-bc78a273-cfca-4a3f-a94c-6e61cf6d6fee.png)

gustn3965 commented:
아, 추가적으로 최근항목 제외한 나머지 앨범들은 날짜가 오름차순으로 나오는 것 같습니다. 의도한게 맞는지 확인해주세요~

jwonyLee committed:
[#141] ref: 피드백 반영
- 앨범 순서 변경
- ContentsCollectionViewCell.imageView.contentMode = .scaleAspectFill
- iOS 14, 14.5에서 첫번째 셀 이미지가 나타나지 않는 현상 수정
- PhotosViewController 탭 바 레이클 크기가 달라지는 현상 수정
  - 320 크기의 디바이스에서 짤리는 현상 수정
- SwiftLint에 맞게 수정
- 파일 정렬

gustn3965 commented on ThingLog/ViewController/WriteViewController+setup.swift:
```diff
@@ -53,6 +29,29 @@ extension WriteViewController {
+        NSLayoutConstraint.activate([
+            doneButton.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+            doneButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+            doneButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
```
safeArea 바텀에 할당해서 지금 화면이  아래와 같이 나오고 있습니다.
![image](https://user-images.githubusercontent.com/48749182/139527141-a4b54e47-3847-4212-a940-d2c2fde7e0c9.png)

gustn3965 commented on ThingLog/ViewController/PhotosViewController+CollectionView.swift:
```diff
@@ -0,0 +1,92 @@
+                options.isSynchronous = true
+                asset.toImage(targetSize: PHImageManagerMaximumSize, options: options) { [weak self] image in
+                    guard let self = self, let image = image else { return }
+                    self.selectedImages.append(image)
```
`WriteViewController` - `WriteImageTableCell`에서 썸내일정도의 화질로만 보여주면 충분 할 것 같은데, 
이 코드 때문인지 필요이상의 고화질의 이미지가 보여지고 있습니다. 이것 때문에 잠깐 랙이 발생하는 것 같기두 하구요.

gustn3965 commented on ThingLog/View/TableVeiwCell/WriteImageTableCell.swift:
```diff
@@ -88,18 +153,31 @@ extension WriteImageTableCell: UICollectionViewDataSource {
+            cell.configure(image: image)
+
+            cell.closeButtonDidTappedCallback = { [weak self] in
+                self?.thumbnailImages.remove(at: indexPath.row - 1)
```
최대한 많이 사진들을 추가하 다음에 맨끝에서부터 삭제나 중간에 삭제하면 앞에 셀들이 안보이는 현상이 있습니다.
디버깅으로 봤을 떄는 아예 셀이 없어지더군요! 

저도 도움을 드리려고 계속 디버깅 해봤는데 이유를 잘 모르겠습니다. ㅠ 
- 정 해결 못할 것 같으면 `reloadSection`을 통해서 아예 0번째 셀 위치로 하는 차선책도 있는 것 같습니다. 
![image](https://user-images.githubusercontent.com/48749182/139527678-6fe81fce-b2c5-430d-8b74-2970a65f149a.png)

gustn3965 commented on ThingLog/View/TableVeiwCell/ImageWithVerticalTwoLabelTableCell.swift:
```diff
@@ -0,0 +1,97 @@
+        return label
+    }()
+
+    private let imageWidth: CGFloat = 80.0
+    private let imageHeight: CGFloat = 100.0
```
피그마에서는 80, 79.5 이던데  ( 아마 80,80 을 의도한 것 같아요 ) 왜 여기서는 80,100 이죠??

jwonyLee replied:
앨범 목록을 보여주기 위해서 테이블뷰를 선택했는데 테이블뷰는 셀 사이 간격을 지정해주는 방법이 없어서 `contentView.frame`에 `Inset`을 추가하는 방식을 사용했는데요. 이렇게 하니, 이미지 뷰가 추가한 `Inset`만큼 작아져서 수치를 조정하게 되었습니다. 이 부분은 컬렉션 뷰로 수정해서 올리겠습니다~

gustn3965 replied:
콜렉션뷰료 꼭 수정하지 않아도돼요! 패딩 추가하는건 다양한 방식으로 할 수 있으니까요, 그냥 궁금해서 물어봤습니다. 저는 주로 셀안에 패딩뷰 넣어서 해서요 ㅎㅎ

jwonyLee replied:
테이블뷰에 `Inset`을 추가하니까 빈 여백을 터치했을 때도 셀이 선택된 것으로 처리가 되는데 이 부분이 사용자가 원치 않는 터치를 했을 때도 동작된다는 걸 알게되서 변경했습니당. 코드 부분에서는 크게 달라진 게 없습니당

gustn3965 commented on ThingLog/View/TableVeiwCell/WriteImageTableCell.swift:
```diff
@@ -73,13 +81,70 @@ extension WriteImageTableCell {
+    /// - Parameter completion: 권한이 승인된 경우 true, 권한이 거부당한 경우 false를 반환한다.
+    private func requestPhotoLibraryAccess(completion: @escaping (Bool) -> Void) {
+        PHPhotoLibrary.requestAuthorization { status in
+            switch status {
```
이건 따로 객체를 만들어서 하는게 좋을 것 같아요. 제 생각에는 지금 뷰에서 많은 일을 하고 있는 듯 합니다.

gustn3965 commented on ThingLog/ViewController/AlbumsViewController.swift:
```diff
@@ -0,0 +1,188 @@
+    private var sections: [AlbumSectionType] = [.all, .favorites, .userCollections]
+    private var allPhotos: PHFetchResult<PHAsset> = PHFetchResult<PHAsset>()
+    private var favorites: PHFetchResult<PHAssetCollection> = PHFetchResult<PHAssetCollection>()
+    private var userCollections: PHFetchResult<PHAssetCollection> = PHFetchResult<PHAssetCollection>()
```
실제 화면에서는 섹션이 없고 셀만으로 구성되어있다고 느껴지고,  테이블뷰가 섹션별로 나뉘었다는 느낌을 받지 못하기 때문에, 섹션이 있다는 것에 다소 혼란스러워요. 일반적으로 섹션별로 나누면 셀들은 섹션별로 그룹이 지어지어서 나타나는게 제일 일반적이니까요.  테이블뷰를 어떻게 구성하는지에 대해 따로 설명을 추가해주시면 좋을 것 같습니다.  글쓰기 화면에서도 물건 이름, 가격 ,구매처가 하나의 섹션으로 분리되어 제가 질문드렸던 것 처럼요.

gustn3965 commented on ThingLog/ViewController/AlbumsViewController.swift:
```diff
@@ -0,0 +1,188 @@
+        switch sectionType {
+        case .all:
+            allPhotos.firstObject?.toImage(targetSize: thumbnailSize, options: options) { [weak self] image in
+                cell.configure(image: image, title: "최근 항목", description: "\(self?.allPhotos.count ?? 0)")
```
바로 아래코드에서는 파라미터를 줄로 잘 나누어주셨는데, 통일해주시면 좋겠습니다. ( 유동적으로 길다 싶으면 줄로 나누는게 좋은 것 같아요 ) 
```swift
            firstAsset.toImage(targetSize: thumbnailSize, options: options) { image in
                cell.configure(image: image,
                               title: collection.localizedTitle ?? "",
                               description: "\(fetchedAssets.count)")
            }
```

gustn3965 commented on ThingLog/ViewController/ContentsCollection/ContentsCollectionViewCell.swift:
```diff
@@ -172,6 +184,13 @@ class ContentsCollectionViewCell: UICollectionViewCell {
                                        endPoint: CGPoint(x: 0.0, y: 0.0))
     }
+
+    private func bindCheckButton() {
+        checkButton.rx.tap
+            .bind { [weak self] in
+                self?.didTappedCheckButtonCallback?()
+            }.disposed(by: disposeBag)
+    }
```
어차피 Rx를 이용해서 바인딩 할 거면  이코드를 사용하지 않고 외부에서 한번에 바인딩하는게 낫지 않나요?

jwonyLee replied:
제가 처음에 외부에서 말씀하신 방법대로 작성했었는데, 셀이 재활용되면서 선택하지 않은 셀이 추가되는 문제가 발생해서 위와 같은 방식으로 작성했습니다. ;(

gustn3965 replied:
;; 이거는 지원님이 저번에 물어보셨는뎅   재사용하니, `prepareForResuse`를 이용해야한다고 저번에 말씀드렸던 것 같은데..

gustn3965 commented on ThingLog/ViewController/PhotosViewController+Albums.swift:
```diff
@@ -0,0 +1,36 @@
+            transition.type = .push
+            transition.subtype = .fromBottom
+            self.albumsViewController.view.layer.add(transition, forKey: "showAlbums")
+            self.albumsViewController.view.isHidden = false
```
오, 하나 배워갑니다!👍

gustn3965 commented on ThingLog/ViewController/ContentsCollection/ContentsCollectionViewCell.swift:
```diff
@@ -81,13 +85,17 @@ class ContentsCollectionViewCell: UICollectionViewCell {
     }()
-    
+
+    var representedAssetIdentifier: String = ""
```
해당 프로퍼티의 필요성을 아래의 `PhotosViewController`를 보고 알게 됐는데요,  그래도 해당 프로퍼티가 필요한 이유를 명시해주면 좋을 것 같습니다.

gustn3965 commented on ThingLog/ViewController/PhotosViewController+CollectionView.swift:
```diff
@@ -0,0 +1,92 @@
+        let asset: PHAsset = assets.object(at: indexPath.item - 1)
+
+        cell.representedAssetIdentifier = asset.localIdentifier
+        asset.toImage(targetSize: thumbnailSize) { image in
```
`toImage`는 `PHImageManager`의 `requestImage`를 래핑하는 메소드이고, 이 메서드는 기본적으로 비동기적으로 작동한다라고 되어있습니다. 그래서 굉장히 많은 사진이 있을때, 이를 빠르게 스크롤 하게 되면 이 메소드들이 다 메모리에 남아서 비동기적으로 모두다 실행되어지고 있습니다. 그래서 이미지가 굉장히 늦게 로딩되는 걸 볼 수 있습니다. 또 디버그에서 CPU만 보더라도 계속해서 처리하고 있다는 걸 보여주듯 CPU가 계속 높음을 유지하고 있습니다. 그래서 일반적으로 이렇게 비동기적으로 작동하는 기능들은 화면구현을 넘어서 빠르게 스크롤도 해보는 테스트도 필요합니다. 

또 하나의 문제점은 위와 같이 진행하다보면, 다시 첫화면으로 돌아오면 첫번째 셀이 카메라셀이 아닌 경우가 발생합니다. 


제가 찾아본 결과,`PHImageManager`는 `requestImage`를 취소하는 메서드가 존재하고 이를 잘 활용하시면 빠르게 로딩될 것입니다.  첫번째 셀 문제는 indexPath를 이용하면 해결 될것 같습니다.

gustn3965 commented on ThingLog/ViewController/PhotosViewController+CollectionView.swift:
```diff
@@ -0,0 +1,92 @@
+        }
+        cell.setupImageViewWithCheckButton()
+        cell.updateCheckButton(string: "", backgroundColor: .clear)
+        DispatchQueue.main.async {
+            self.updateSelectedOrder()
+        }
+    }
```
이미 메인큐에서 진행되고 있을 텐데 한번더 main async를 한 이유는 무엇인가요?? 

그리고, 하나의 셀마다 updateSeelctedOrder를 호출할 필요가 있나요? 해당 메서드는 모든 선택된셀들을 다 확인하면서 업데이트하는 메소드 같은데, 하나의 셀에서는 하나의 셀만 확인하면 안되나요?

gustn3965 commented on ThingLog/ViewController/PhotosViewController.swift:
```diff
@@ -0,0 +1,303 @@
+            }.disposed(by: disposeBag)
+    }
+
+    private func bindNotification() {
```
선택된앨범을 받아온다는 내용이 포함된 메서드명과 주석이 있으면 좋을 것 같습니다.

gustn3965 commented on ThingLog/ViewController/PhotosViewController.swift:
```diff
@@ -0,0 +1,303 @@
+extension PhotosViewController {
+    /// `selectedIndexPath`을 초기화한다.
+    func resetSelectedIndexPath() {
+        selectedIndexPath.forEach { indexPath in
+            guard let cell: ContentsCollectionViewCell = collectionView.cellForItem(at: indexPath) as? ContentsCollectionViewCell else {
+                return
+            }
+
+            cell.updateCheckButton(string: "", backgroundColor: .clear)
+            cell.layoutIfNeeded()
+        }
+
+        selectedIndexPath = []
```
아래와 같이 간단하게 해도 될 것 같은데, 어떻게 생각하시나요? 
```swift
        selectedIndexPath = []
        collectionView.reloadData()
```

gustn3965 commented on ThingLog/ViewController/PhotosViewController.swift:
```diff
@@ -0,0 +1,303 @@
+
+    /// 네비게이션 우측 상단에 있는 확인 버튼에 선택한 개수를 변경한다. 선택한 항목이 없으면 "확인"으로 변경한다.
+    private func updateSelectedCountLabel() {
+        guard !selectedIndexPath.isEmpty else {
```
지원님께서 `if selectedIndexPath.isEmpty` 가 더 읽기 쉽다고 하셨던 것 같은데,  지켜주세요

gustn3965 commented on ThingLog/Extension/PHAsset+.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
+
+extension PHAsset {
+    func toImage(targetSize size: CGSize, contentMode: PHImageContentMode = .aspectFill, options: PHImageRequestOptions? = nil, completion: @escaping (UIImage?) -> Void) {
```
파라미터가 길 경우 줄로 나누도록 하는게 좋을 것 같아요~!

gustn3965 reviewed (commented):
후..리뷰만 하는데 3시간 이상 걸린 것 같네요. 이건 제가 그렇게 하라고 했으니 괜찮습니다. 다음부터는 너무 길면 힘들 것 같아요 ㅎㅎ.. 

코멘트 확인 부탁드립니다 ~ 

그리고 `**ViewController+***` 와 같이 파일이 여러개 생긴다면, 폴더로 묶는건 어떨까요?? 저도 `LoginViewController`같이 폴더로 묶도록 할게요!

jwonyLee committed:
[#141] ref: 피드백 반영
- iOS 14 에서 썸네일 이미지를 삭제할 때 일부 셀이 안 보이는 문제 해결 (reloadSection)
- PhotosViewController.setupContentsCell() 에서 updateSelectedOrder() 호출 대신 현재 셀에 대한 부분만 업데이트
- PhotosViewController 에서 checkButton을 탭할 때마다 asset을 동기적으로 가져오는 부분 수정
- bindNotification -> bindNotificationSelectAlbum
- doneButton.bottomAnchor 변경
- ContentsCollectionViewCell.checkButton bind를 외부에서 하게 변경
- requestPhotoAccess 관련 메서드를 UserPermissionsViewModel로 분리
- NotificationCenter.passSelectImage를 passSelectAssets로 변경, 전달하는 객체를 PHAsset으로 변경
  - WriteViewController에서 해당 알림을 구독하면서 thumbnailSize를 fetch
  - WriteViewModel에서 해당 알림을 구독하면서 originalImage를 fetch, 비동기로 동작

jwonyLee committed:
[#141] ref: PHCachingImageManager를 이용해 이미지 캐싱
Loading and Caching Assets and Thumbnails 참고

jwonyLee commented:
피드백 주신 내용을 토대로 수정했습니다.
- 코드 Indent 맞추기
- Notification.Name
  - `passSelectImage` → `passSelectAssets`로 변경
- WriteVC
  - iOS 14에서 중간 혹은 가장 끝에 있는 썸네일을 삭제하면 보이는 셀 중 앞에 있는 셀이 안 보이는 문제 해결 → `collectionView.reloadSection()` 사용
  - `doneButton.bottomAnchor` 변경
- PhotosVC
  -  컬렉션 뷰에서 셀을 그릴 때 `updateSelectedOrder()` 호출 대신 현재 셀에 대한 부분만 업데이트
  - `cell.checkButton`을 탭할 때마다 asset을 가져와서 저장하는 코드 제거
  - `bindNotification()` → `bindNotificationSelectAlbum()`
  - `ContentsCollectionViewCell.checkButton`의 Callback 클로저 제거, bind를 외부에서 하는 것으로 변경
  - `Notification.passSelectAssets` 전달 객체를 `[PHAsset]`으로 변경 (기존: passSelectImage, [UIImage])
  - `PHCachingImageManager`를 이용해 보여주는 이미지 캐싱
    - 참고: https://developer.apple.com/documentation/photokit/loading_and_caching_assets_and_thumbnails
- WriteImageTableCell
  - 사진 권한 메서드를 `UserPermissionsViewModel`로 분리
- WriteViewModel
  - `Notification.passSelectAssets`를 구독하면서 전달받은 `[PHAsset]`객체를 originalSize로 변환해서 저장

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -32,8 +35,63 @@ final class WriteViewModel {
+                self?.executeOriginalImages(with: assets)
+            }.disposed(by: disposeBag)
+    }
+
```
여기 주석 내용이 빠졌어요~

gustn3965 commented on ThingLog/ViewController/Photos/PhotosViewController+Caching.swift:
```diff
@@ -0,0 +1,77 @@
+
+        // 캐싱을 시작하고 정지할 assets 계산
+        // indexPath.item이 0인 경우는 카메라 셀을 표시하므로 필터링
+        let (addedRects, removedRects): ([CGRect], [CGRect]) = differencesBetweenRects(previousPreheatRect, preheatRect)
+        let addedAssets: [PHAsset] = addedRects
+            .flatMap { rect in collectionView.indexPathsForElements(in: rect) }
+            .filter { indexPath in indexPath.item == 0 }
+            .map { indexPath in assets.object(at: indexPath.item) }
+        let removedAssets: [PHAsset] = removedRects
+            .flatMap { rect in collectionView.indexPathsForElements(in: rect) }
+            .filter { indexPath in indexPath.item == 0 }
+            .map { indexPath in assets.object(at: indexPath.item) }
+
+        // PHCachingImageManager가 캐싱한 자산을 업데이트
+        imageManager.startCachingImages(for: addedAssets,
+                                        targetSize: thumbnailSize, contentMode: .aspectFill, options: nil)
+        imageManager.stopCachingImages(for: removedAssets,
+                                       targetSize: thumbnailSize, contentMode: .aspectFill, options: nil)
+        // 향후 비교를 위해 계산된 사각형을 저장
+        previousPreheatRect = preheatRect
```
음, 저는 이게 제대로 캐싱이 되어지고있는지 잘 모르겠습니다. print로 찍어보면 스크롤을 계속 해도 캐싱할 Asset과 삭제할 Asset이 전혀 나타나지 않고 있거든요, 
![image](https://user-images.githubusercontent.com/48749182/140022851-e69abe8b-7643-4183-9062-9343bf22623d.png)

gustn3965 reviewed (commented):
나머지 코드들은 다  괜찮은 것 같습니다.  우선적으로 코멘트들을 달았습니다.


그외, 지원님이 보시다시피 굉장히 의존성이 여러개 있기 때문에, 어디서 어디로 데이터가 옮겨가는지, 등을 대략적으로라도 정리해놓는게 앞으로도 좋을 것 같아요! ( 예를 들어, 사진데이터는 어떤 뷰컨트롤러에서, 어떤 뷰모델 또는 어떤 뷰컨트롤러, 어떤 뷰로 옮겨간다 이런식으로요, ) 왜냐면 나중에 코드보면 코드로만 이해하는데 조금 시간이 걸릴 수도 있거든요. 저는 복잡한 로직관해서는 대략적으로 정리하면서 코드를 짜는 편이이에요.

 
 -------------- 
 수정하신 이미지 캐싱에 대해서 제가 스크린샷 찍어봤습니다
기존 코드를 돌렸을때, ( 제 디바이스) 굉장히 빠르게 스크롤 하여 멈추게 되면, 계속해서 CPU가 유지하고있고, 사진도 조금 늦게 업데이트 됩니다.  ( 처음에는 흐릿하게 보여지다가, 나중에 선명해져요 ) 
![지원](https://user-images.githubusercontent.com/48749182/140024595-116a937d-45a2-468e-ad80-d81568ad4551.gif)
.

제가 코드를 수정하고서 돌렸을떄,  빠르게 스크롤하게 되면 잠깐 CPU가 높아졌다가 금방 사라지는 걸 볼 수 있습니다. 사진도 금방 업데이트 되구요. 
![수정](https://user-images.githubusercontent.com/48749182/140024705-ef796ab5-63f6-4b3a-bd90-db8cdcc28a3e.gif)

jwonyLee committed:
[#141] ref: 피드백 반영
- 누락된 주석 추가
- 이미지 캐싱 필터링 오류 수정
- 셀이 재사용될 때 ImageRequest.cancle 추가 (현수님)

jwonyLee commented:
- 누락된 주석 추가했습니다.
- 캐싱하는 코드에서 제가 필터링을 잘못 하고 있던 부분이 있어서 수정했습니다. (--)(__)
<img width="1046" alt="스크린샷 2021-11-04 오후 7 35 10" src="https://user-images.githubusercontent.com/15073405/140299290-78d83e81-0ee1-44d7-b25c-d0f48f6fc1be.png">

- 더불어 현수님이 수정하셨던 셀이 재활용될 때 image request를 취소하는 코드도 포함하였습니다.

gustn3965 commented on ThingLog/ViewController/Photos/PhotosViewController.swift:
```diff
@@ -0,0 +1,306 @@
+            collectionView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+            collectionView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
+            collectionView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+            collectionView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
```
`safeAreaLayoutGuide`의 `botomAnchor`로 지정하고 있어서 노치있는 디바이스에서는 `PhotosViewController`가 살짝 뜬 상태로 나타납니다.

gustn3965 commented on ThingLog/ViewController/Photos/PhotosViewController+CollectionView.swift:
```diff
@@ -0,0 +1,104 @@
+            if cell.representedAssetIdentifier == asset.localIdentifier {
+                cell.update(image: image)
+            }
+        }
```
저번 리뷰에서도 말씀드렸었는데, 간혹 아래로 많이 스크롤한다음에 가장 첫번째로 다시 가게 되면, 첫번째셀이 카메라가 아닌경우가 종종 있습니다.  indexPath를 이용해서 해결하거나 다른 방법을 이용하셔야 할 것 같습니다.

gustn3965 reviewed (approved):
고생하셨습니다. 아래 코맨트들 확인하시고 테스트하신다음에 머지하시면 될 것 같아요. 

그리고 이전 풀리퀘에서 `ContentsCollectionViewCell`의 `checkButton`이 `CheckView`로 변경되었는데, 이 과정에서 `UIControl`을 상속받는 `UIImageView`만 포함하는 뷰로 변경됐습니다. 제가 여기서 카운팅하는 것을 캐치하지 못했습니다. 죄송합니다! 그래서 머지하시면 에러가 발생할텐데, 유동적으로 `CheckView`에 `버튼`이나 `레이블`을 추가하시면 될 것 같습니다. 🙏🏻

jwonyLee committed:
[#141] ref: 피드백 반영
- collectionView BottomAnchor 수정
- imageManager.requestImage 에서 indexPath.item == 0인 경우 return

jwonyLee committed:
merge: upstream develop branch

jwonyLee committed:
Merge branch 'develop' into view/141-photosVC

jwonyLee committed:
ref: code incorrectly during merge
